src/TriggerBinding/SqlTableChangeMonitor.cs

```diff
@@ -0,0 +1,769 @@
+                        try
+                        {
+                            // Update the version number stored in the global state table if necessary before using it.
+                            using (SqlCommand updateTablesPreInvocationCommand = this.BuildUpdateTablesPreInvocation(connection, transaction))
+                            {
+                                await updateTablesPreInvocationCommand.ExecuteNonQueryAsync(token);
+                            }
+
+                            // Use the version number to query for new changes.
+                            using (SqlCommand getChangesCommand = this.BuildGetChangesCommand(connection, transaction))
+                            {
+                                var rows = new List<IReadOnlyDictionary<string, string>>();
+                                using (SqlDataReader reader = await getChangesCommand.ExecuteReaderAsync(token))
+                                {
+                                    while (await reader.ReadAsync(token))
+                                    {
+                                        rows.Add(SqlBindingUtilities.BuildDictionaryFromSqlRow(reader));
+                                    }
+                                }
+
+                                this._rows = rows;
+                            }
+
+                            // If changes were found, acquire leases on them.
+                            if (this._rows.Count > 0)
+                            {
+                                using (SqlCommand acquireLeasesCommand = this.BuildAcquireLeasesCommand(connection, transaction))
+                                {
+                                    await acquireLeasesCommand.ExecuteNonQueryAsync(token);
+                                }
+                            }
+                            transaction.Commit();
+                        }
+                        catch (Exception ex)
+                        {
+                            this._logger.LogError("Commit Exception Type: {0}", ex.GetType());
+                            this._logger.LogError("  Message: {0}", ex.Message);
+                            try
+                            {
+                                transaction.Rollback();
+                            }
+                            catch (Exception ex2)
+                            {
+                                this._logger.LogError("Rollback Exception Type: {0}", ex2.GetType());
+                                this._logger.LogError("  Message: {0}", ex2.Message);
+                            }
+                        }
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                // If there's an exception in any part of the process, we want to clear all of our data in memory and
+                // retry checking for changes again.
+                this._rows = new List<IReadOnlyDictionary<string, string>>();
```

Comment:
Is there a reason for re-initializing the list rather than just clearing them?

Reply:
I will need to recheck on this to confirm. Since `this._rows` is shared among both concurrent loops, if the other loop is processing the list referenced by `_rows`, it would be safer to not touch that list by instead start with a fresh one.

Reply:
This is correct. I have deliberately defined `this._rows` to be `IReadOnlyList<IReadOnlyDictionary<string, string>>` so we can only call read-only method on the list pointed by it. We cannot modify the list by reassign new list to `this._rows`. This ensures the updating the list is an atomic operation.